File: 2022/09/07/keyboardRow.js
/**
 * https://leetcode.com/problems/max-consecutive-ones/
 * @author Andrew Jarombek
 * @since 9/7/2022
 */

const findWords = function(words) {
    const rows = ["qwertyuiop", "asdfghjkl", "zxcvbnm"];
    const sets = [];

    for (let row of rows) {
        const set = new Set();

        for (let c of row) {
            set.add(c);
        }

        sets.push(set);
    }

    const result = [];

    for (let word of words) {
        let set = null;
        let c = word.charAt(0).toLowerCase();

        if (sets[0].has(c)) {
            set = sets[0];
        } else if (sets[1].has(c)) {
            set = sets[1];
        } else {
            set = sets[2];
        }

        let matches = true;
        for (let c of word) {
            if (!set.has(c.toLowerCase())) {
                matches = false;
                break;
            }
        }

        if (matches)
            result.push(word);
    }

    return result;
};

let result = findWords(["abc", "def", "ghi", "jkl"]);
let expected = ["jkl"]
console.assert(expected.every((value, index) => value === result[index]));

File: 2023/07/06/go/keyboard_row.go
/**
 * https://leetcode.com/problems/keyboard-row/
 * Author: Andrew Jarombek
 * Date: 7/6/2023
 */

package _go

import "strings"

func findWords(words []string) []string {
	rows := make(map[byte]int)

	for _, c := range "qwertyuiop" {
		rows[byte(c)] = 1
	}

	for _, c := range "asdfghjkl" {
		rows[byte(c)] = 2
	}

	for _, c := range "zxcvbnm" {
		rows[byte(c)] = 3
	}

	var result []string

	for _, word := range words {
		lower := strings.ToLower(word)
		row := rows[lower[0]]
		i := 1
		valid := true

		for valid && i < len(lower) {
			if row != rows[lower[i]] {
				valid = false
			} else {
				i++
			}
		}

		if valid {
			result = append(result, word)
		}
	}

	return result
}
